tempodb/encoding/v1/data_reader.go

```diff
@@ -39,29 +41,29 @@ func NewNestedDataReader(r common.DataReader, encoding backend.Encoding) (common
 // Read returns the pages requested in the passed records.  It
 // assumes that if there are multiple records they are ordered
 // and contiguous
-func (r *dataReader) Read(ctx context.Context, records []*common.Record) ([][]byte, error) {
-	compressedPages, err := r.dataReader.Read(ctx, records)
+func (r *dataReader) Read(ctx context.Context, records []*common.Record, buffer []byte) ([][]byte, []byte, error) {
+	compressedPages, buffer, err := r.dataReader.Read(ctx, records, buffer)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	// now decompress
 	decompressedPages := make([][]byte, 0, len(compressedPages))
 	for _, page := range compressedPages {
 		reader, err := r.getCompressedReader(page)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 
-		page, err := ioutil.ReadAll(reader)
+		page, err := tempo_io.ReadAllWithEstimate(reader, len(page)+1) // +1 prevents an extra alloc on when encoding=None
```

Comment:
Why does encoding=None need the +1 ?

Reply:
if len exactly matches what you want to read then line 629 hits:

https://golang.org/src/io/io.go?s=19249:19287#L629

and allocates more even those the next `r.Read()` is going to return `io.EOF`